code/datums/storage/storage.dm

```diff
@@ -1117,6 +1091,116 @@ GLOBAL_LIST_EMPTY(cached_storage_typecaches)
+	corner4.icon_state = "storage_corner_bottomright"
+	corner2.update_appearance()
+	corner3.update_appearance()
+	corner4.update_appearance()
+
+	rowjoin2.icon_state = "storage_rowjoin_right"
+
+	storage_sets[ui_style] = list(cells, corner1, corner2, corner3, corner4, rowjoin1, rowjoin2)
+
+/// Updates views of all objects in storage and stretches UI to appropriate size
+/datum/storage/proc/orient_storage()
+	var/adjusted_contents = real_location.contents.len
+	var/list/datum/numbered_display/numbered_contents
+	if(numerical_stacking)
+		numbered_contents = process_numerical_display()
+		adjusted_contents = numbered_contents.len
+
+	//if the ammount of contents reaches some multiplier of the final column (and its not the last slot), let the player view an additional row
+	var/additional_row = (!(adjusted_contents % screen_max_columns) && adjusted_contents < max_slots)
+
+	var/columns = clamp(max_slots, 1, screen_max_columns)
+	var/rows = clamp(CEILING(adjusted_contents / columns, 1) + additional_row, 1, screen_max_rows)
+
+	for (var/ui_theme in storage_sets)
+		var/list/atom/movable/screen/storage/ui_set = storage_sets[ui_theme]
+		ui_set[STORAGE_CELL_INDEX].screen_loc = "[screen_start_x]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y] to [screen_start_x + columns - 1]:[screen_pixel_x],[screen_start_y + rows - 1]:[screen_pixel_y]"
+
+		ui_set[STORAGE_TOP_LEFT_CORNER_INDEX].screen_loc = "[screen_start_x + 1]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y] to [screen_start_x + columns - 1]:[screen_pixel_x],[screen_start_y + rows - 1]:[screen_pixel_y]"
+		ui_set[STORAGE_TOP_RIGHT_CORNER_INDEX].screen_loc = "[screen_start_x]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y] to [screen_start_x + max(0, columns - 2)]:[screen_pixel_x],[screen_start_y + rows - 1]:[screen_pixel_y]"
+		ui_set[STORAGE_BOTTOM_LEFT_CORNER_INDEX].screen_loc = "[screen_start_x + 1]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y] to [screen_start_x + columns - 1]:[screen_pixel_x],[screen_start_y + rows - 1]:[screen_pixel_y]"
+		ui_set[STORAGE_BOTTOM_RIGHT_CORNER_INDEX].screen_loc = "[screen_start_x]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y] to [screen_start_x + max(0, columns - 2)]:[screen_pixel_x],[screen_start_y + rows - 1]:[screen_pixel_y]"
+
+		ui_set[STORAGE_LEFT_ROWJOIN_INDEX].screen_loc = "[screen_start_x]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y + 27] to [screen_start_x]:[screen_pixel_x],[screen_start_y + max(0, rows - 2)]:[screen_pixel_y + 27]"
+		ui_set[STORAGE_LEFT_ROWJOIN_INDEX].alpha = (rows > 1) * 255
+
+		ui_set[STORAGE_RIGHT_ROWJOIN_INDEX].screen_loc = "[screen_start_x + columns - 1]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y + 27] to [screen_start_x + columns - 1]:[screen_pixel_x],[screen_start_y + max(0, rows - 2)]:[screen_pixel_y + 27]"
+		ui_set[STORAGE_RIGHT_ROWJOIN_INDEX].alpha = (rows > 1) * 255
+
+		storage_closers[ui_theme].screen_loc = "[screen_start_x + columns]:[screen_pixel_x - 5],[screen_start_y]:[screen_pixel_y]"
```

Comment:
This similarly makes me want to jump off a building. are we essentially building a set of overlays here or what.

Reply:
yeah, exactly. I expanded on how original UI did it, i.e. have a single object and stretch it to make it tile.